File: data-structure-and-algorithm/cpp/permutations-ii.cpp
// 47. Permutations II
// Difficulty: Medium

// Given a collection of numbers that might contain duplicates, return all possible unique permutations.

// For example,
// [1,1,2] have the following unique permutations:

// [
//   [1,1,2],
//   [1,2,1],
//   [2,1,1]
// ]

// Time : O(n!)
// Space: O(n)

class Solution {
public:
  vector<vector<int>> permute(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    vector<bool> visited(nums.size(), false);
    vector<vector<int>> ans;
    vector<int> path;
    permute(nums, visited, path, ans);

    return ans;
  }

private:
  void permute(vector<int>& nums, vector<bool>& visited, vector<int>& path, vector<vector<int>>& ans) {
    if (nums.size() == path.size()) {
      ans.push_back(path);
      return;
    }

    for (int i = 0; i < nums.size(); i++) {
      if (!visited[i]) {
        path.push_back(nums[i]);
        visited[i] = true;
        permute(nums, visited, path, ans);
        path.pop_back();
        visited[i] = false;
        for (; i < nums.size() - 1 && nums[i] == nums[i + 1]; i++);
      }
    }
  }
};

File: data-structure-and-algorithm/java/permutations-ii.java
/**
 * Permutations II
 * 
 * Given a collection of numbers that might contain duplicates, return all possible unique permutations.
 * 
 * For example,
 * [1,1,2] have the following unique permutations:
 * [1,1,2], [1,2,1], and [2,1,1].
 */

import java.util.Arrays;

public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        if (nums == null) {
            throw new NullPointerException();
        }

        List<list<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<Integer>();
        boolean[] visited = new int[nums.length];

        Arrays.sort(nums);
        permuteHelper(nums, visited, 0, path, result);

        return result;
    }

    public void permuteHelper(int[] nums, boolean[] visited, int start, List<Integer> path, List<List<Integer>> result) {
        if (start == nums.length) {
            result.add(path);
        } else {
            for (int i = start; i < nums.length; ++i) {
                if (visited[i] == false) {
                    visited[i] = true;
                    path.add(nums[i]);
                    permuteHelper(nums, visited, i + 1, path, result);
                    path.remove(path.size() - 1);
                    visited[i] = false;
                    for (; i < nums.length - 1 && nums[i] == nums[i + 1]; ++i);
                }
            }
        }
    }
}
